server/backup_runner.go

```diff
@@ -51,13 +58,69 @@ func (s *BackupRunner) Run(ctx context.Context, wg *sync.WaitGroup) {
 					}
 				}()
 				s.startAutoBackups(ctx, runningTasks, &mu)
+				s.downloadBinlogFiles(ctx)
 			}()
 		case <-ctx.Done(): // if cancel() execute
+			s.downloadBinlogWg.Wait()
 			return
 		}
 	}
 }
 
+func (s *BackupRunner) downloadBinlogFiles(ctx context.Context) {
+	instanceList, err := s.server.store.FindInstanceWithDatabaseBackupEnabled(ctx, db.MySQL)
+	if err != nil {
+		log.Error("Failed to retrieve MySQL instance list with at least one database backup enabled", zap.Error(err))
+		return
+	}
+
+	for _, instance := range instanceList {
+		s.downloadBinlogMu.Lock()
+		if _, ok := s.downloadBinlogTasks[instance.ID]; !ok {
+			s.downloadBinlogTasks[instance.ID] = true
+			go s.downloadBinlogFilesForInstance(ctx, instance, s.server.profile.DataDir, s.server.mysqlutil)
+			s.downloadBinlogWg.Add(1)
+		}
+		s.downloadBinlogMu.Unlock()
+	}
+}
+
+func (s *BackupRunner) downloadBinlogFilesForInstance(ctx context.Context, instance *api.Instance, dataDir string, mysqlutil mysqlutil.Instance) {
+	log.Debug("Downloading binlog files for MySQL instance", zap.String("instance", instance.Name))
+	defer func() {
+		s.downloadBinlogMu.Lock()
+		delete(s.downloadBinlogTasks, instance.ID)
+		s.downloadBinlogMu.Unlock()
+		s.downloadBinlogWg.Done()
+	}()
+	driver, err := getAdminDatabaseDriver(ctx, instance, "", "" /* pgInstanceDir */)
+	if err != nil {
+		if common.ErrorCode(err) == common.DbConnectionFailure {
+			log.Warn("Cannot connect to instance", zap.String("instance", instance.Name), zap.Error(err))
+			return
+		}
+		log.Error("Failed to get driver for MySQL instance when downloading binlog", zap.String("instance", instance.Name), zap.Error(err))
+		return
+	}
+	defer driver.Close(ctx)
+
+	binlogDir := getBinlogAbsDir(dataDir, instance.ID)
+	if err := createBinlogDir(dataDir, instance.ID); err != nil {
+		log.Error("Failed to create binlog directory", zap.Error(err))
+		return
+	}
+	mysqlDriver, ok := driver.(*mysql.Driver)
+	if !ok {
+		log.Error("Failed to cast driver to mysql.Driver", zap.String("instance", instance.Name))
+		return
+	}
+	mysqlDriver.SetUpForPITR(mysqlutil, binlogDir)
+	if err := mysqlDriver.FetchAllBinlogFiles(ctx); err != nil {
```

Comment:
Are we downloading the last and incomplete one? This might take a lot of IO.

Reply:
Yes and yes.
I'll skip downloading the last binlog file here, and only download it when doing PITR.